File: backend/app/models/feedback.py
"""
Feedback/Support Ticket models for FitCheck AI.
"""
from datetime import datetime
from enum import Enum
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, Field, EmailStr


class TicketCategory(str, Enum):
    """Support ticket categories."""
    BUG_REPORT = "bug_report"
    FEATURE_REQUEST = "feature_request"
    GENERAL_FEEDBACK = "general_feedback"
    SUPPORT_REQUEST = "support_request"


class TicketStatus(str, Enum):
    """Ticket status values."""
    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    CLOSED = "closed"


class DeviceInfo(BaseModel):
    """Device/browser information for context."""
    platform: Optional[str] = None  # 'web', 'ios', 'android'
    os_version: Optional[str] = None
    device_model: Optional[str] = None
    browser: Optional[str] = None
    screen_size: Optional[str] = None


class CreateFeedbackRequest(BaseModel):
    """Request to create a new feedback/support ticket."""
    category: TicketCategory
    subject: str = Field(..., min_length=3, max_length=200)
    description: str = Field(..., min_length=10, max_length=5000)
    contact_email: Optional[EmailStr] = None  # For anonymous submissions
    device_info: Optional[DeviceInfo] = None
    app_version: Optional[str] = Field(None, max_length=50)
    app_platform: Optional[str] = Field(None, max_length=20)  # 'web', 'ios', 'android'


class FeedbackResponse(BaseModel):
    """Response after creating feedback."""
    id: UUID
    category: TicketCategory
    subject: str
    status: TicketStatus
    created_at: datetime
    message: str = "Thank you for your feedback!"

    class Config:
        from_attributes = True


class TicketListItem(BaseModel):
    """Ticket in list view."""
    id: UUID
    category: TicketCategory
    subject: str
    status: TicketStatus
    created_at: datetime


class TicketListResponse(BaseModel):
    """List of user's tickets."""
    tickets: List[TicketListItem]
    total: int

File: frontend/src/api/feedback.ts
/**
 * Feedback API endpoints
 */

import { apiClient, getApiError } from './client'
import type { ApiEnvelope } from '../types'

export interface DeviceInfo {
  platform: string
  os_version?: string
  device_model?: string
  browser?: string
  screen_size?: string
}

export type TicketCategory = 'bug_report' | 'feature_request' | 'general_feedback' | 'support_request'
export type TicketStatus = 'open' | 'in_progress' | 'resolved' | 'closed'

export interface FeedbackResponse {
  id: string
  category: TicketCategory
  subject: string
  status: TicketStatus
  created_at: string
  message: string
}

export interface TicketListItem {
  id: string
  category: TicketCategory
  subject: string
  status: TicketStatus
  created_at: string
}

export interface TicketListResponse {
  tickets: TicketListItem[]
  total: number
}

/**
 * Submit feedback with optional attachments
 */
export async function submitFeedback(params: {
  category: TicketCategory
  subject: string
  description: string
  contactEmail?: string
  attachments?: File[]
  deviceInfo?: DeviceInfo
  appVersion?: string
}): Promise<FeedbackResponse> {
  try {
    const formData = new FormData()
    formData.append('category', params.category)
    formData.append('subject', params.subject)
    formData.append('description', params.description)

    if (params.contactEmail) {
      formData.append('contact_email', params.contactEmail)
    }

    if (params.deviceInfo) {
      formData.append('device_info', JSON.stringify(params.deviceInfo))
    }

    if (params.appVersion) {
      formData.append('app_version', params.appVersion)
    }

    formData.append('app_platform', 'web')

    if (params.attachments) {
      for (const file of params.attachments) {
        formData.append('attachments', file)
      }
    }

    const response = await apiClient.post<ApiEnvelope<FeedbackResponse>>(
      '/api/v1/feedback',
      formData,
      {
        headers: { 'Content-Type': 'multipart/form-data' },
      }
    )
    return response.data.data
  } catch (error) {
    throw getApiError(error)
  }
}

/**
 * Get current user's tickets
 */
export async function getMyTickets(
  limit: number = 20,
  offset: number = 0
): Promise<TicketListResponse> {
  try {
    const response = await apiClient.get<ApiEnvelope<TicketListResponse>>(
      '/api/v1/feedback/my-tickets',
      { params: { limit, offset } }
    )
    return response.data.data
  } catch (error) {
    throw getApiError(error)
  }
}
